### DISCUSSION TITLE
C# Lowering

### DISCUSSION THREAD

**Discussion by giscus** [linkdotnet/Blog.Discussions#42] [General]
# C# Lowering

Did you ever hear the word "compiler magic" or "syntactic sugar"? Probably yes and therefore we want to dissect what this "magic" really is!
We can see how we can predict performance or bugs by "lowering" our code. Also we will see how things like foreach, var, lock, using, async, await, yield, anonymous types, record, stackalloc, pattern matching, Blazor components, deconstructor, extension methods... do not really exist.

https://steven-giesel.com/blogPost/69dc05d1-9c8a-4002-9d0a-faf4d2375bce

**Comment by mrjohnsonit** [comment]
Curious where the term 'Lowering' came from? If Syntactic Sugar == Lowering why rebrand?

**Reply by linkdotnet** [reply]
Hmmm good question, lowering makes more sense to be honest :D
As we are going from high level features to **low** level features...also in the sense of closer to the underlying hardware / os.

**Reply by naine** [reply]
"Syntactic Sugar" is a more general term to describe language features that don't enable you to do anything you can't do without them, just allowing you to do things in a cleaner way with less code. "Lowering" is an implementation technique within the compiler that involves transforming the code internally before compiling it. Syntactic sugar features aren't necessarily implemented using a lowering (but often are).

**Reply by linkdotnet** [reply]
Very fair point and absolutely right!

**Reply by MSBassSinger** [reply]
> "Syntactic Sugar" is a more general term to describe language features that don't enable you to do anything you can't do without them, just allowing you to do things in a cleaner way with less code. "Lowering" is an implementation technique within the compiler that involves transforming the code internally before compiling it. Syntactic sugar features aren't necessarily implemented using a lowering (but often are).

Two ways of saying the same thing.  "Lowering" is just a made up term, albeit an inaccurate one, for doing things the hard way.  In most cases, there is no demonstrable advantage in taking the Luddite approach.  If it compiles to the same IL, all you succeeded in doing is making your C# code harder to read.

**Reply by naine** [reply]
I think you're confusing coding style with the inner workings of the compiler. Lowered code is not meant to be readable, it's internal to the compiler. It also doesn't exactly apply to C# code, since it's applied to the parsed syntax tree.

And lowering is not a made up term, but it's a compiler concept not a programming concept.

**Reply by MSBassSinger** [reply]
> I think you're confusing coding style with the inner workings of the compiler. Lowered code is not meant to be readable, it's internal to the compiler. It also doesn't exactly apply to C# code, since it's applied to the parsed syntax tree.
> 
> And lowering is not a made up term, but it's a compiler concept not a programming concept.

No, not confusing anything.  "Lowering" has relevance at the compiler level.  In reading the comments, some readers seem to find value in applying the concept in their C# code, which is the point I was addressing.  Those coming from a Java background would be tempted to go down that rabbit hole.

 Thus, at the C# coding level, it is a made-up term to describe the opposite of syntactic sugar, and poor coding practice when used without providing a net positive value.

For the reader who may want more information on lowering in the compiler:

https://mattwarren.org/2017/05/25/Lowering-in-the-C-Compiler/

https://shiver.github.io/post/csharp-lowering/

https://subscription.packtpub.com/book/application-development/9781787286832/9/ch09lvl1sec66/implementing-lowering-code-generation-support-for-a-new-c-language-feature

**Comment by ltcmelo** [comment]
Here's a library that rewrites certain C# constructs: https://github.com/ShiftLeftSecurity/SharpSyntaxRewriter

**Reply by linkdotnet** [reply]
Thanks for the reference! it really looks interesting. What was your initial spark to start this @ltcmelo ?

**Reply by ltcmelo** [reply]
For certain constructs, it's more convenient to implement an static analysis if they are in lowered/desugared form.

**Comment by MSBassSinger** [comment]
I highly recommend knowing what the IL looks like for various spoonfuls of syntactic sugar.  

For example, I do that, and that is why I do not use "using", but use try-catch-finally, which gives better exception control and does not swallow the using object's constructor exception.

But, if the syntactic sugar and the "lowering" process yield the same IL, then just use the syntactic sugar, as there is no advantage to "lowering".

**Reply by smurthys** [reply]
```
I do not use "using", but use try-catch-finally, which gives better exception control and does not swallow the using object's constructor exception.
```

Good point. For exactly the reason you specify, `using` is syntactic sugar only for a specific pattern of using try-catch-finally. I'd use it regardless of whether it is "lowered" if I need that exact semantics, just to improve readability.

**Comment by fmargueirat** [comment]
I'm wondering why is the array copied in the 2nd example? Is that to ignore changes that may be done in the original array in multi-threading code?

**Reply by linkdotnet** [reply]
Good question, I wanted to expand the article about that topic, but I am not 100% sure (but roughly 90%).

My educated guess is that the `int[] array2 = array` part is to be thread-safe. Keep in mind that `foreach` is a very general construct that can be used everywhere.
There is a chance that while you enumerate through your enumeration, the pointer may change (because of a second thread that runs in parallel). If you copy the pointer beforehand, your `foreach` is safe against that.

**Reply by naine** [reply]
The language rules dictate that the expression in the foreach statement is evaluated exactly once. Imagine if you had `foreach (var x in GetArray())`, the method must be called exactly once with the returned array reference assigned to a temporary variable. Even if the expression is a simple local variable reference (being a local, it is not accessible by other threads), it is possible to reassign the variable within the loop, but the loop must continue over the array it referred to when entering the loop, hence the compiler must copy the array reference to a temporary. Of course the compiler could optimize out the temporary in the specific example shown, but this would complicate the implementation so it leaves the optimization to the JIT.

**Reply by linkdotnet** [reply]
Thanks for the input - I will update that passage (and, of course giving credits you).

**Comment by Caltor** [comment]
Why is everyone so obsessed about syntactic sugar? All control structures beyond _if_ and _while_ are syntactic sugar. You could write every program in them. That is the essence of turing completeness.

**Comment by toupswork** [comment]
How is `stackalloc` syntactic sugar?

**Reply by linkdotnet** [reply]
There is a local rewriter for the `stackalloc` keyword: https://github.com/dotnet/roslyn/blob/main/src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StackAlloc.cs

**Reply by linkdotnet** [reply]
That was my basis for labeling as lowered statement

**Reply by toupswork** [reply]
oh ok. Cause I don't see any lowering when I try it on Sharplab

**Reply by linkdotnet** [reply]
That's fair - and my assumption also was that one should see it in sharplab.io.
I am not 100% sure when those Rewriters are invoked. Well normally as part of the compiler chain, but I also would suspect a difference in the sharplab output

**Reply by toupswork** [reply]
I found a little trick today that I thought was pretty neat. Whenever we call a repo, service, or Api, we return a `Task<ServiceResult<T>>`. Service result holds just an enum to indicate the status of the call (success, fail, etc) and the value returned T. Say you are often needing to make multiple calls and use implicit parallelism with Task.WhenAll(), and you want to create a Fluent API that allows you to add a number of `Func<Task<ServiceResult<T>>>` to the fluent api to chain them together, but the T is different for each one you add. But you need to keep track of each T because at the end,  you want to call and await an async method that runs the Tasks in parallel, and the response object contains a method that has output parameters of  `ServiceResult<T>` for each Func you added. How might you implement it?

So something like this
```
var result = await AsyncHelper.StartWith(()=>repo.Get<Customers>())
     .ThenAdd(()=> apiClient.GetClient("5000"))
     .ThenAdd(()=>payrollApi.GetCostCenter(4325))
     // ... more
     .WhenAll();

result.GetAll(out var customers, out var client, out var costCenter);
```

**Reply by linkdotnet** [reply]
Probably you need something that holds a list of your tasks or the execution of your tasks: `List<Func<Task<T>>>`.
With every `ThenAdd` you would add an entry to that (as well as with `StartsWith`.

Then you can easily await that via `Task.WhenAll(myList.Select(l => (Task)l));`.
Not sure about you `GetAll` without declaring something like `GetAll<List<Customer>, Client, CostCenter>`.

**Reply by toupswork** [reply]
I had to find a very clever way to do it. I didn't even think it could be done. I've tried before, but to no success. At one point I thought what if I cast them as an object and use Unsafe.As() to cast them back. I'll send you the code later. The thing is that C# generics is not like Java's. It enforces it so you can only use the actual types.

**Reply by linkdotnet** [reply]
Yeah Java is basically "everything is object".
But yes that would be very interesting to see - especially because in my lifetime as dev, I used `Unsafe.As` exactly once and once only.

**Reply by linkdotnet** [reply]
@toupswork 

Here is my version without any `Unsafe` code:

```csharp
var result = await TaskHelper.StartWith(() => Task.FromResult(10))
    .Then(() => Task.FromResult("Hello World"))
    .Then(() => Task.FromResult(10d))
    .WaitAll();

result.GetAll(out int intResult, out string stringResult, out double doubleResult);

public class TaskHelper
{
    private readonly List<Func<Task<object>>> _tasks = [];

    private TaskHelper() { }

    public static TaskHelper StartWith<T>(Func<Task<T>> task)
    {
        var helper = new TaskHelper();
        helper._tasks.Add(() => task().ContinueWith(t => (object)t.Result));
        return helper;
    }

    public TaskHelper Then<T>(Func<Task<T>> task)
    {
        _tasks.Add(() => task().ContinueWith(t => (object)t.Result));
        return this;
    }

    public async Task<Result> WaitAll()
    {
        var tasks = _tasks.Select(task => task()).ToArray();
        var results = await Task.WhenAll(tasks);
        return new Result(results);
    }
}

public class Result(IReadOnlyList<object> results)
{
    public void GetAll<T>(out T result)
    {
        ArgumentOutOfRangeException.ThrowIfLessThan(results.Count, 1);
        result = (T)results[0];
    }
    
    public void GetAll<T1, T2>(out T1 result1, out T2 result2)
    {
        ArgumentOutOfRangeException.ThrowIfLessThan(results.Count, 2);
        result1 = (T1)results[0];
        result2 = (T2)results[1];
    }
    
    public void GetAll<T1, T2, T3>(out T1 result1, out T2 result2, out T3 result3)
    {
        ArgumentOutOfRangeException.ThrowIfLessThan(results.Count, 3);
        result1 = (T1)results[0];
        result2 = (T2)results[1];
        result3 = (T3)results[2];
    }
}
```

**Reply by toupswork** [reply]
You're on the right track. Doing it with `Task<T>` is easier than `Task<ServiceResult<T>>`. The hard part is that you're taking in a `Func<Task<ServiceResult<T>>>` but the generic signature of the type is either a series of `T` or `ServiceResut<T>`, but not the Func or Task. So externally, the type is a higher level than what was passed in.  Then it has to convert the internal Funcs and retype them based on the series of `<T1,T2,T3>` etc. My first iteration was to use an interface of `IAsyncPipe<T>` and then `T` would become a `Tuple<T1,T2,T3>` and the underlying class would store them as an `ITuple` then the all the methods were written using extension methods, since it's easier to do then doing it on the underlying class. But I think I found another way, but I got kind of busy because we have a production release coming up, so once things slow down, I'll go back to it.

**Reply by toupswork** [reply]
Also it looks like your forcing the consumer to have to remember and enter the generic types in the right order. For mine, it's inferred automatically

**Reply by linkdotnet** [reply]
Yeah I do force the user to know (in the same order of calling) which type they expect.

I can't wrap my head around how to infer that automatically without the use of Source Code Generators.

My approach would look not very different with `Task<ServiceResult<T>>` - basically just one more layer, but wouldn't change anything implementation-wise.

**Reply by toupswork** [reply]
Here take a look at where I am so far. 

```C#
public readonly record struct ServiceResult<TResult>(ResultStatus Status, TResult? Result)
{
  // omitted implementation
}

public readonly record struct ServiceResultSet<TResult>(ResultStatus Status, List<TResult> Result)
{
  // omitted implementation
}


public interface IApiChain<T,TResult>
{
  IApiChain<T,T2, TResult, ServiceResultSet<T2>> Add<T2>(Func<Task<ServiceResultSet<T2>>> func);
  IApiChain<T, T2, TResult, ServiceResult<T2>> Add<T2>(Func<Task<ServiceResult<T2>>> func);
}
public interface IApiChain<T1, T2, TResult1, TResult2> 
{
  IApiChain<T1, T2, T3, TResult1, TResult2, ServiceResultSet<T3>> Add<T3>(Func<Task<ServiceResultSet<T3>>> func);
  IApiChain<T1, T2, T3, TResult1, TResult2, ServiceResult<T3>> Add<T3>(Func<Task<ServiceResult<T3>>> func);
  Task<IChainResult<T1, T2, TResult1, TResult2>> AwaitAllAsync();
}

public interface IApiChain<T1, T2, T3, TResult1, TResult2, TResult3> 
{
  IApiChain<T1, T2, T3, T4, TResult1, TResult2, TResult3, ServiceResultSet<T4>> Add<T4>(Func<Task<ServiceResultSet<T4>>> func);
  IApiChain<T1, T2, T3, T4, TResult1, TResult2, TResult3, ServiceResult<T4>> Add<T4>(Func<Task<ServiceResult<T4>>> func);
  Task<IChainResult<T1, T2, T3, TResult1, TResult2,TResult3>> AwaitAllAsync();
}
public interface IApiChain<T1, T2, T3, T4, TResult1, TResult2, TResult3, TResult4>
{
  Task<IChainResult<T1, T2, T3, T4, TResult1, TResult2, TResult3, TResult4>> AwaitAllAsync();
}
public interface IChainResult<T1, T2, TResult1, TResult2>
{
  void Deconstruct(out TResult1 result1, out TResult2 result2);
  void GetValues(out T1 value1, out T2 value2);
}
public interface IChainResult<T1, T2, T3, TResult1, TResult2, TResult3>
{
  void GetValues(out T1 result1, out T2 result2, out T3 result3);
  void Deconstruct(out TResult1 result1, out TResult2 result2, out TResult3 result3);
}
public interface IChainResult<T1, T2, T3, T4, TResult1, TResult2, TResult3, TResult4>
{
  void GetValues(out T1 result1, out T2 result2, out T3 result3, out T4 result4);
  void Deconstruct(out TResult1 result1, out TResult2 result2, out TResult3 result3, out TResult4 result4);
}
public static class ApiChain
{
  public static IApiChain<T, ServiceResult<T>> StartWith<T>(Func<Task<ServiceResult<T>>> first) => 
    new ApiChain<T, ServiceResult<T>>(first);

  public static IApiChain<T, ServiceResultSet<T>> StartWith<T>(Func<Task<ServiceResultSet<T>>> first) => 
    new ApiChain<T, ServiceResultSet<T>>(first);
}

file sealed class ApiChain<T,TResult>(Func<Task<TResult>> func): IApiChain<T, TResult>
{
  public IApiChain<T, T2, TResult, ServiceResultSet<T2>> Add<T2>(Func<Task<ServiceResultSet<T2>>> func)
  {throw new NotImplementedException();}

  public IApiChain<T, T2, TResult, ServiceResult<T2>> Add<T2>(Func<Task<ServiceResult<T2>>> func)
  {throw new NotImplementedException();}
}

file sealed class ApiChain<T1, T2, TResult1, TResult2>(Func<Task<T1>> func1, Func<Task<T2>> func2): IApiChain<T1, T2, TResult1, TResult2>
{
  public IApiChain<T1, T2, T3, TResult1, TResult2, ServiceResultSet<T3>> Add<T3>(Func<Task<ServiceResultSet<T3>>> func)
  {throw new NotImplementedException();}

  public IApiChain<T1, T2, T3, TResult1, TResult2, ServiceResult<T3>> Add<T3>(Func<Task<ServiceResult<T3>>> func)
  {throw new NotImplementedException();}

  public Task<IChainResult<T1, T2, TResult1, TResult2>> AwaitAllAsync()
  {throw new NotImplementedException();}
}

file class ChainResult<T1, T2, TResult1, TResult2>(TResult1 result1, TResult2 result2) : IChainResult<T1, T2, TResult1, TResult2>
{
  public void Deconstruct(out TResult1 first, out TResult2 second)
  {
    first = result1;
    second = result2;
  }

  public void GetValues(out T1 value1, out T2 value2)
  {throw new NotImplementedException();}
}
```
![image](https://github.com/linkdotnet/Blog.Discussions/assets/127423854/0507ad3d-10eb-4c52-a072-ee5581eac8a0)

**Reply by toupswork** [reply]
Notice, I have two Service Result types, and it's able to mix and match them and preserve the type information, and it's able to unwrap the underlying values from both `Task<T>` and `ServiceResult<T>`. Also, because the Result types are structs, there's no boxing. So it maintains performance without needing to force cast it from more abstract to concrete

**Reply by linkdotnet** [reply]
I guess I do have a version that is a bit simpler - I will come back soon (maybe melt this into its own blog post).

**Reply by linkdotnet** [reply]
Okay - have a working version with: https://steven-giesel.com/blogPost/d55cdae5-7294-40cc-a9d5-7552e092205e/fluent-api-to-await-multiple-calls-and-get-their-respective-results

@toupswork

**Reply by linkdotnet** [reply]
Adding `ServiceResult` shouldn't much more code to the solution IMHO

**Reply by toupswork** [reply]
That's great and very concise.  I guess you're trying to one up me :) 
The difference with mine is that it's not meant to be general purpose; it needs to restrict to only ServiceResult or ServiceResultSet, because it is going to check to ensure that all of them returned a success, instead of the consumer having to check each one individually. That's a requirement I didn't tell you about. If one of them fails, they all have to fail, and I have take the first non-success status and return it.  I also prefer to use interfaces when working with Fluent APIs. That's just a preference to reduce information leak, and the classes are scoped to the file.  Yours is more general purpose for all types. But that means you have to down cast everything to an object, so it's going to box value types.  So to prevent boxing, and to check the result status of each ServiceResult or ServiceResultSet to ensure it was a success, then yes I do need more code.

**Reply by toupswork** [reply]
If you want to get into a competition over succinctness, then I'll grab my first iteration which used only one class and generic type and kept track of the underlying types by setting the one generic type to a Tuple—so there was no T1, T2…TN—and then stored each func internally as an ITuple that it then iterated like an array. :)

**Reply by linkdotnet** [reply]
It sounds a bit like structured concurrency what you want to achieve with that.

And yes, mine is a very general solution to the specific problem at hand. While boxing might be an issue here, I don't think someone would use such a construct in a high-performance scenario anyway. 
Anyhow - quite interesting problem!

**Reply by toupswork** [reply]
Well someone who might want something like this (me and my team) is a person writing an API that has to constantly aggregate data from other APIs and Databases.  All APIs do (if designed right) is async I/O operations and offload heavy computation tasks to microservices. So in an API that does nothing but constant async I/O, that kind of is a hot path. There is overhead to box and unbox, and then there can be boxing going into the TaskHelper and boxing coming out, along with boxing when it calls the Task.WaitAll.  Also for reference types, there is a small overhead to go from abstract to concrete because it has to perform type verification. So it can all add up and put pressure on the GC, fragment the heap, increase collection cycles that causes all threads to pause (depending on GC mode and number of cores this can vary) while it cleans up memory. So no requests are being serviced when the GC pauses threads, and pausing threads has its own overhead as well by the way. And this is especially important when an API runs in a container with limited resources. Microsoft considered it a big enough issue that they created logging source generators to prevent boxing when logging. And my result types are value types to lower heap allocations. And Microsoft has added more value types, such as Value Tuples and Value Tasks. So for demonstration purposes, that's a cool trick you have. But real life is a bit more complicated.

**Reply by linkdotnet** [reply]
I understand the need for such hot paths, but I would argue that they should be evaluated case by case. In the majority of cases, boxing/unboxing will not make any difference (also mainly because the majority of DTO's are classes). 
Also given the lambdas `ThenAdd(() => ...)` can create GC pressure and lead to subtle bugs and long-living objects one did not intend.

Often, you will go much further if you optimize other levels before asking yourself whether or not something gets boxed. Again that really depends on the concrete situation/application/scenario.

**Reply by toupswork** [reply]
The JIT can optimize boxing in simple scenarios, not something like this. And to your point about lambdas, I know better than to create a closure, so that's not an issue

**Reply by toupswork** [reply]
Also, everywhere I've worked, repos and API clients don't just return a DTO; they return a result object. And those result types are often value types like what I have.

**Reply by linkdotnet** [reply]
That makes sense and that is why I said, that it has to be justified case by case.

In the average case of my projects complex DTO's were always `class`es (if it wasn't a `bool`, `int` and so on).
The `ServiceResult` I saw until now was also always a `class`.

**Reply by toupswork** [reply]
I have benchmarks that show it's faster with a struct, even with value copying. Besides the memory overhead of the method table, Sync Block, padding alignment, and garbage collection just to create a small class with two fields, the JIT can pass the value type through registers rather than pushing it onto the stack.
